downstream.go

```diff
@@ -154,12 +150,48 @@ func (dc *downstreamConn) upstream() *upstreamConn {
 	return upstream
 }
 
+func (dc *downstreamConn) marshalEntity(uc *upstreamConn, name string) string {
+	for _, r := range name {
+		switch r {
+		case '#', '&', '+', '!':
+			return dc.marshalChannel(uc, name)
+		}
+		break
+	}
+	return dc.marshalNick(uc, name)
+}
+
+func (dc *downstreamConn) marshalChannel(uc *upstreamConn, name string) string {
+	if dc.upstream() != nil {
+		return name
+	}
+	return name + "/" + uc.network.GetName()
+}
+
 func (dc *downstreamConn) unmarshalChannel(name string) (*upstreamConn, string, error) {
 	if uc := dc.upstream(); uc != nil {
 		return uc, name, nil
 	}
 
-	// TODO: extract network name from channel name if dc.upstream == nil
+	network := ""
+	if i := strings.LastIndexAny(name, "/@"); i >= 0 {
+		network = name[i+1:]
+	}
+	if i := strings.IndexAny(name, "/@"); i >= 0 {
```

Comment:
I think we want to use `LastIndexAny` in all cases, because we want `#thing/other-thing/network` to be split into channel name `#thing/other-thing` and network name `network`. The logic is different for usernames because we want to allow things like `user@client-id/network` for instance (`client-id` allows to properly send history to multiple clients).